### PR TITLE
[issue-139] Improve documentation

### PR DESCRIPTION
**Pull request by rsearls**
Please review this major rewrite and restructuring of ./README.md 
and provisioners/README.md and provide feedback particularly on 
correctness of the information. 

Part of #139

### CODE REVIEW

**Comment by fabiobrz** [comment]
Thank you @rsearls - this is, simply put, great.

It'll take some time to review the changes, but such an improvement will add a huge value, and therefore I bound the PR to the release milestone.

File: provisioners/README.md [hunk 1]
```diff
@@ -1,204 +1,264 @@
...
+	public SpyProvisioner getProvisioner(Application application) {
+		if (SpyApplication.class.isAssignableFrom(application.getClass()))
+			return new SpyProvisioner((SpyApplication) application);
+		return null;
+	}
+}
 ```
+
+### Step 4 Register the ProvisionerFactory as a service
+Edit service file, _./provisioners/src/main/resources/META-INF/services/org.jboss.intersmash.provision.ProvisionerFactory_ and add _org.secret.agents.SpyProvisionerFactory_ and rebuilt Intersmash.
```
Comment (fabiobrz) [inline provisioners/README.md:207]: s/rebuilt/rebuild/

File: provisioners/README.md [hunk 2]
```diff
@@ -1,204 +1,264 @@
...
+
+
+### Step 5 Provide a demonstration application
+The last task is to write and add a simple application that demonstrates the service working in directory, examples.
```
Comment (fabiobrz) [inline provisioners/README.md:211]: I'd s/directory, examples/the `examples` directory/

**Review by fabiobrz — CHANGES_REQUESTED** [review]
Thanks again @rsearls for this PR.
This is a huge and valuable contribution to the project, so kudos to your commitment.

Most of my comments are just about typos or minor things, like formatting to have a better rendering. A few of them are about the contents.

Feel free to let me know what you think.

File: README.md [hunk 3]
```diff
@@ -1,93 +1,128 @@
...
+It helps the user prototype and test complex interoperability scenarios on
+kubernetes compliant cloud-native environments and platforms, most notably OpenShift.  (*Other Kubernetes implementations will be supported in the future.*)
+
+Intersmash is designed following some of the same principles as Arquillian.  It is integrated with JUnit.
+It handles all the plumbing (i.e. provisioning)
+of the cloud container. It deploys the test scenario *product componets*,
+manages their life-cycle orchestration, and
+undeploys and cleans up the environment on test completion.
+*(The term services and product components will be used interchangeably.)*
+It is an extensible
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:14]: done

File: README.md [hunk 4]
```diff
@@ -1,93 +1,128 @@
...
+framework.  If there is a cloud service not yet implemented, the user can easily
+create an implementation and integrate it into the framework.
+The framework is cloud implementation agnostic.  It will be possible *(in the future)*
+to test on different Kubernetes compliant cloud implementations.  This will
+ensure application portability.
+
+## Framework Design
+
+Intersmash provides a set of annotations, APIs, and service provisioners
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:23]: done

File: README.md [hunk 5]
```diff
@@ -1,93 +1,128 @@
...
+the developer uses to describe in the test class, the test scenario.
+This information is used by the framework to deploy the services and to manage their
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:25]: Kept previous edit and current 2nd sentence.  I my opinion it reads well and is clear.

File: README.md [hunk 6]
```diff
@@ -1,93 +1,128 @@
...
+life-cycle orchestration directly from the test class.
+A curated list of service "provisioners" is provided for the user's convenience,
+for example WildflyImageOpenShift, Kafka Operator, PostgreSQLTemplateOpenShift.
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:28]: done

File: README.md [hunk 7]
```diff
@@ -1,93 +1,128 @@
...
+The full list of provided services can be found [here](./docs/Supported-Product-Components.md).
+
+
+[XTF](https://github.com/xtf-cz/xtf) is used by Intersmash to communicate
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:32]: done

File: README.md [hunk 8]
```diff
@@ -1,93 +1,128 @@
...
+with OpenShift.  This tool provides a means to pass configuration properties
+from Intersmash to OpenShift.
+Intersmash also uses class, [XTFConfig](https://github.com/xtf-cz/xtf/blob/master/core/src/main/java/cz/xtf/core/config/XTFConfig.java),
+internally.
+XTF supports four ways to provide properties. In priority order they are,
+
+  - **System properties**
+  - **Environment variables**
+  - **test.properties file**: residing in root of the project designed to contain
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:41]: done

File: README.md [hunk 9]
```diff
@@ -1,93 +1,128 @@
...
+    user specific setup.  A different location of the "test.properties" file
+    can be specified using system property, **-Dxtf.test_properties.path**.
+  - **global-test.properties file**: residing in root of the project designed to contain
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:44]: done

File: README.md [hunk 10]
```diff
@@ -1,93 +1,128 @@
...
+    a shared setup. You can use  **-Dxtf.global_test_properties.path** property
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:45]: done

File: README.md [hunk 11]
```diff
@@ -1,93 +1,128 @@
...
+
+A table of properties Intersmash uses to configure and execute tests can be
+found [here](./docs/XTF-Configuration-Properties.md).
+
+
+There are three main components that make up the framework:
+
+* **annotations**: Intersmash annotations are used to declare the
+  *product components* that make up the test scenario.
+  These statements are declared in the test class.
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:56]: done

File: README.md [hunk 12]
```diff
@@ -1,93 +1,128 @@
...
+
+
+There are three main components that make up the framework:
+
+* **annotations**: Intersmash annotations are used to declare the
+  *product components* that make up the test scenario.
+  These statements are declared in the test class. 
+
+  There are four annotations
+  - **@Intersmash**: contains a list of identified services that make up the test scenario.
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:59]: done

File: README.md [hunk 13]
```diff
@@ -1,93 +1,128 @@
...
+  - **@Service**:  the class reference to an individual service
+  - **@ServiceUrl**: injects into the test class as a String or URL the address of a service
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:61]: done

File: README.md [hunk 14]
```diff
@@ -1,93 +1,128 @@
...
+  - **@ServiceProvisioner**: injects into the test class a provisioner that enables control over platform-specific actions or example, scaling the deployment up and down.
+
+
+* **applications**: An Intersmash provided interface class that describes
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:65]: done

File: README.md [hunk 15]
```diff
@@ -1,93 +1,128 @@
...
+  the configuration and specific properties of a given service.  For every supported *product component*
+  there is a corresponding "application" interface class.  The Framework's naming convention for these classes is *ProductComponentName*Application.java. (e.g. BootableJarOpenShiftApplication.java, ActiveMQOperatorApplication.java HelmChartOpenShiftApplication.java, ... etc)
+  The interface provides default values for most of the service's configuration information.
+  The user is required to create an implementation class of the interface and
+  provide some select information.  It is this user created implemented class that is declared in the annotations, @Service, @ServiceUrl, and @ServiceProvisioner.
+
+
+* **provisioners**: Classes that takes care of deploying, undeploying, managing the
+  life-cycle orchestration of services, and controls the specific workflow that
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:74]: done

File: README.md [hunk 16]
```diff
@@ -1,93 +1,128 @@
...
-    <groupId>org.jboss.intersmash</groupId>
-    <artifactId>intersmash-core</artifactId>
-  </dependency>
-  <dependency>
-      <groupId>org.jboss.intersmash</groupId>
-      <artifactId>intersmash-provisioners</artifactId>
-  </dependency>
-</dependencies>
-```
+The following outlines a simple test scenario in which PostgresSql and Wildfly are used.
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:83]: done

File: README.md [hunk 17]
```diff
@@ -1,93 +1,128 @@
...
+    @ServiceProvisioner(PostgresqlApp.class) 
+    private OpenShiftProvisioner postgresqlProvisioner;
 
     @Test
     public void test() {
         // Do your thing.
     }
 }
 ```
+1. Using Intersmash annotations the user registers their interface implementation
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:108]: done

File: README.md [hunk 18]
```diff
@@ -1,93 +1,128 @@
...
+   classes of the services to be used in this test.
+2. Injects into the test a string containing the URL to communicate with Wildfly.
+3. Injects into the test the provisioner for Wildfly and PostgreSQL.  These classes
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:111]: done

File: README.md [hunk 19]
```diff
@@ -103,179 +138,95 @@ public class PostgresqlApplication implements PostgreSQLTemplateOpenShiftApplica
...
+  }
+
+  @Override
+  public BuildInput getBuildInput() {
+    return new BuildInputBuilder().archive(app).build();
+  }
+}
+```
+
+## Maven Archives
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:160]: done

File: README.md [hunk 20]
```diff
@@ -103,179 +138,95 @@ public class PostgresqlApplication implements PostgreSQLTemplateOpenShiftApplica
...
+Two maven archives are needed to build and run with Intersmash.
+```xml
+<dependencies>
+  <!-- contains the Intersmash annotation definitions -->
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:164]: done

File: README.md [hunk 21]
```diff
@@ -103,179 +138,95 @@ public class PostgresqlApplication implements PostgreSQLTemplateOpenShiftApplica
...
 * **core** - core interfaces, annotations and extensions.
 * **provisioners** - implementations of interfaces from `core` for selected services.
-* **testsuite** - tests that verify the integration with OpenShift, hence a cluster is needed.  
-  * **test-eployments** - sources for shared deployments which are used by the testsuite.
-    Moreover, the `intersmash-deployments-provider` submodule provides support to get path to compiled deployments. Or
-    you can get any deployment by GAV.
-* **examples** - examples that illustrate how to integrate Intersmash.
+* **testsuite** - tests that verify the integration with OpenShift. (**Note:** an OpenShift cluster is needed in order to run them.)
+  * **test-deployments** - sources for shared deployments which are used by the testsuite.
+    * **deployments-provider** - provides support to get path to compiled deployments or
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:210]: done

File: README.md [hunk 22]
```diff
@@ -103,179 +138,95 @@ public class PostgresqlApplication implements PostgreSQLTemplateOpenShiftApplica
...
-* intersmash-provisioners-_&lt;version&gt;_.jar - Actual provisioners implementation
-* intersmash-provisioners-_&lt;version&gt;_-sources.jar - Unzip to get the sources or use in your IDE
-* intersmash-provisioners-_&lt;version&gt;_-javadoc.jar - unzip to get Javadoc
-
-## Publishing Intersmash Library artifacts
-
-The Intersmash Library project is configured to publish some artifacts to the public Maven repository, so that external 
-projects can access them, e.g.: the `intersmash-core.jar` artifact which contains the `@Intersmash` annotation 
-definition.
+## Platforms
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md:214]: done

File: docs/Supported-Product-Components.md [hunk 23]
```diff
@@ -0,0 +1,30 @@
+
+### Supported Product Components
```
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/Supported-Product-Components.md:2]: done

File: docs/Supported-Product-Components.md [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+
+| Product Component             | Interface Class Name                   |
```
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/Supported-Product-Components.md:5]: done

File: docs/Supported-Product-Components.md [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+|-------------------------------|----------------------------------------|
+| **Application Server Images** |                                        |
+| BootableJarOpenShift          | BootableJarOpenShiftApplication        |
+| WildflyImageOpenShift         | WildflyImageOpenShiftApplication       |
+| Eap7ImageOpenShift            | Eap7ImageOpenShiftApplication          |
+| **Database Images**           |                                        |
+| MysqlImageOpenShift           | MysqlImageOpenShift Application        |
```
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/Supported-Product-Components.md:12]: done

File: docs/Supported-Product-Components.md [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+| PostgreSQLImageOpenShift      | PostgreSQLImageOpenShiftApplication    |
+| **Operator services**         |                                        |
+| ActiveMQ Operator             | ActiveMQOperatorApplication            |
+| Hyperfoil Operator            | HyperfoilOperatorApplication           |
+| Infinispan Operator           | InfinispanOperatorApplication          |
+| Kafka Operator                | KafkaOperatorApplication               |
+| Keycloak Operator             | KeycloakOperatorApplication            |
+| RhSso Operator                | RhSsoOperatorApplication               |
```
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/Supported-Product-Components.md:20]: done

File: docs/Supported-Product-Components.md [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+| Wildfly Operator              | WildflyOperatorApplication             |
+| **Templated services**        |                                        |
+| Eap7LegacyS2iBuildTemplate    | Eap7LegacyS2iBuildTemplateApplication  |
+| Eap7TemplateOpenShift         | Eap7TemplateOpenShiftApplication       |
+| PostgreSQLTemplateOpenShift   | PostgreSQLTemplateOpenShiftApplication |
+| RhSsoTemplateOpenShift        | RhSsoTemplateOpenShiftApplication      |
+| **Helm Chart**                |                                        |
```
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/Supported-Product-Components.md:27]: done

File: provisioners/README.md [hunk 28]
```diff
@@ -1,204 +1,264 @@
...
+
+- **Automatic**: Automatic (or application dictated) provisioning is a way to delegate the whole provisioning process to the Application class, rather than depending on a provisioner.
+
+
+
+## Adding a new Service
+Here are the steps in adding a new service to Intersmash.
+
+1. Create a new Application interface class which extends from an existing Application interface.
+[UML diagram of Application interface hierarchy](../docs/UML-Application-SPIs-Diagram.jpg).
```
[the remark this answers is not in the document]
Comment (rsearls) [reply provisioners/README.md:72]: done

File: docs/UML-Application-SPIs-Diagram.jpg
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/UML-Application-SPIs-Diagram.jpg:1]: done

File: provisioners/README.md [hunk 29]
```diff
@@ -1,204 +1,264 @@
...
+2. Create a new Provisioner class which implements or extends an existing Provisioner interface or class respectively. [UML diagram of Provisioner interface hierarchy](../docs/UML-Provisioners-SPIs-Diagram.jpg).
+3. Create a new ProvisionerFactory class which implements the _ProvisionerFactory_ interface.
+4. Add your new ProvisionerFactory class to the service file, _./provisioners/src/main/resources/META-INF/services/org.jboss.intersmash.provision.ProvisionerFactory_ 
+5. Provide a demonstration application in the examples directory.
+
+## Provisioning Example
+Here is a simple example that adds a spy service to the framework.
```
[the remark this answers is not in the document]
Comment (rsearls) [reply provisioners/README.md:79]: done

File: provisioners/README.md [hunk 30]
```diff
@@ -1,204 +1,264 @@
...
+import io.fabric8.kubernetes.api.model.EnvVar;
 
-	@Override
-	public PostgreSQLTemplate getTemplate() {
-		return PostgreSQLTemplate.POSTGRESQL_EPHEMERAL;
+public interface SpyApplication extends OpenShiftApplication, HasSecrets,
+    HasEnvVars {
+
+	// Application configuration
+	default List<String> getCliScript() {
```
[the remark this answers is not in the document]
Comment (rsearls) [reply provisioners/README.md:97]: done

File: provisioners/README.md [hunk 31]
```diff
@@ -1,204 +1,264 @@
...
+
+	@Override
+	public void deploy() {
+		deployImage();
+	}
+
+	@Override
+	public void undeploy() {
+		OpenShiftUtils.deleteResourcesWithLabel(openShift, APP_LABEL_KEY, spyApp.getName());
+		// when using git repo S2I process creates some custom maps and build pods
```
[the remark this answers is not in the document]
Comment (rsearls) [reply provisioners/README.md:144]: done

File: provisioners/README.md [hunk 32]
```diff
@@ -1,204 +1,264 @@
...
+            // This is an long task of many steps and thus will not be
+            // listed here.  Look at existing provisioners for examples.
+        }
 }
 ```
 
-These classes can implement their own provisioning process by using the above-mentioned annotations in order to 
-accomplish a declarative configuration, instead that requiring a specific provisioner that would instead leverage 
-a client API.
+### Step 3 Creating a new ProvisionerFactory class
```
[the remark this answers is not in the document]
Comment (rsearls) [reply provisioners/README.md:188]: done

File: provisioners/README.md (hunk 1 above)
Comment (rsearls) [reply provisioners/README.md:207]: done

File: provisioners/README.md (hunk 2 above)
Comment (rsearls) [reply provisioners/README.md:211]: done

File: provisioners/README.md [hunk 33]
```diff
@@ -1,204 +1,264 @@
...
+user credentials, database name and mount path.
+Abstract class, _DBImageOpenShiftProvisioner_ provides basic configuration.
+A provisioner's concrete class provides the extension to these types of services.
+[MysqlImageOpenShiftProvisionerFactory](src/main/java/org/jboss/intersmash/provision/openshift/MysqlImageOpenShiftProvisionerFactory.java)
+provides a reference to a _Provisioner_ and _Application_ implementing this feature.
+
+### Operator-based Provisioning
+Operator services automates the behavior needed to deploy
+a given service on cloud environments via APIs that leverage the
+[Operator Framework](https://github.com/operator-framework).  Operator-based provisioners implement a common contract and high level behavior which is defined by Intersmash's, abstract class, _OperatorProvisioner_.
```
[the remark this answers is not in the document]
Comment (rsearls) [reply provisioners/README.md:235]: done

File: docs/Supported-Product-Components.md
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/Supported-Product-Components.md:1]: done

File: provisioners/README.md [hunk 34]
```diff
@@ -1,204 +1,264 @@
...
-			log.info("READY: " + pod.getStatus().getContainerStatuses().get(0).getReady());
-		});
+package org.secret.agents;
+
+// a long list of imports here
+
+public class SpyProvisioner implements OpenShiftProvisioner<SpyApplication> {
+
+	private final SpyApplication spyApp;
+	private final String CLI_LAUNCH_SCRIPT = "CLI_LAUNCH_SCRIPT";
```
[the remark this answers is not in the document]
Comment (rsearls) [reply provisioners/README.md:122]: done

**Comment by fabiobrz** [comment]
> /retest

Should rebased on main, first.

File: docs/Supported-Product-Components.md
Comment (fabiobrz) [inline docs/Supported-Product-Components.md:1]: Can we remove this and use the other table - i.e. Provisioner-by-Product.md - everywhere?
It has the same information, plus supported product/community variants.

**Review by fabiobrz — CHANGES_REQUESTED** [review]
Hi @rsearls - thanks for addressing my comments.
I've dropped a couple more about minor issues, and one about the two tables which are documenting provisioners, those should be merged into one. 

Fell free to let me know what you think.

File: docs/XTF-Configuration-Properties.md [hunk 35]
```diff
@@ -0,0 +1,89 @@
+
+### Intersmash XTF Configuration Properties
+
+* The following properties can be used to configure Intersmash build and test execution, see the [CI checks e2e tests
+  script](.ci/openshift-ci/build-root/e2e-test-prod.sh) as an example:
```
[the remark this answers is not in the document]
Comment (rsearls) [reply docs/XTF-Configuration-Properties.md]: done

File: README.md [hunk 36]
```diff
@@ -1,93 +1,65 @@
...
+It helps the user prototype and test complex interoperability scenarios on
+kubernetes compliant cloud-native environments and platforms, most notably OpenShift.  (*Other Kubernetes implementations will be supported in the future.*)
 
-If you work as an enterprise application developer, runtime components engineer, quality engineer or in DevOps, 
-Intersmash will help you prototype and test complex interoperability scenarios on cloud-native environments 
-and platforms, most notably on OpenShift (and on Kubernetes as well, soon :) ).
+Intersmash is designed with these principles.  
+* It is integrated with JUnit.
+* It handles all the plumbing (i.e. provisioning) of the cloud container. It deploys the test
+scenario *product componets*, manages their life-cycle orchestration, and
```
[the remark this answers is not in the document]
Comment (rsearls) [reply README.md]: done

File: docs/Supported-Product-Components.md
Comment (rsearls) [reply docs/Supported-Product-Components.md:1]: done

**Review by fabiobrz — APPROVED** [review]
Changes LGTM, I am approving and merging the changes, thanks again @rsearls !

The documentation will be also improved additionally via the changes in #159, where we'll add information about supported versions

**Pull request merged**